Add prepare-commit.sh setup step in how to contribute patches

strk opened the pull request:
Also use real step numbers so you don't need a markdown renderer to read the content

strk committed:
Add pre-commit hooks setup step in how to contribute patches

Also use real step numbers so you don't need a markdown renderer
to read the content and use consistent capitalization in phrases

Suggest to use pre-commit

[skip ci]

benoitdm-oslandia commented on CONTRIBUTING.md:
```diff
@@ -15,9 +15,10 @@ You can help us **by submitting bug reports or fixing bugs** in the [QGIS bug tr
-1. make your changes
-1. commit to your repository
-1. and then [create a pull request](https://help.github.com/articles/creating-a-pull-request-from-a-fork/).
+2. Install the pre-commit hook: `ln -s ../../scripts/prepare_commit.sh .git/hooks/pre-commit`
```
```suggestion
2. Install the pre-commit hook: `pre-commit install --install-hooks`
```
Will do the job by using the `.pre-commit.yaml` file. The yaml file already reference the `scripts/prepare_commit.sh`

strk replied:
I don't have a "pre-commit" command, which package should contain it ?

strk replied:
I'm trying to make contributors life easier, so the least things they need to install, the better

ptitjano replied:
For the record, I think this is a bad idea to not advertise `pre-commit`

nirvn commented:
@strk , is the --install-hooks necessary? It works without that argument for me.

nirvn commented:
@benoitdm-oslandia , right, but what's the difference here between `pre-commit install` and `pre-commit install --install-hooks` practically speak? I'm merely curious here :)

ptitjano commented:
> @benoitdm-oslandia , right, but what's the difference here between `pre-commit install` and `pre-commit install --install-hooks` practically speak? I'm merely curious here :)

See: https://github.com/pre-commit/pre-commit.com/issues/255

Basically, by adding `--install-hooks`, you also install the venv of the different hooks. If you ony do `pre-commit install`, the venvs will be created the first time you run `pre-commit`

strk commented:
I've rebased and updated the hook install command to work with just `git` and supporting worktrees.

I've also tried the `pre-commit` way. Package does exist on Debian 12 and hooks do install, but then the hook doesn't work for me that way:
```
An error has occurred: InvalidManifestError: 
==> File /home/strk/.cache/pre-commit/repop4ji3b_1/.pre-commit-hooks.yaml
==> At Hook(id='clang-format')
==> At key: types_or
==> At index 9
=====> Type tag 'textproto' is not recognized.  Try upgrading identify and pre-commit?
Check the log at /home/strk/.cache/pre-commit/pre-commit.log
```

Is there any good reason for the increased complexity of using this `pre-commit` thing ?

ptitjano commented:
> I've rebased and updated the hook install command to work with just `git` and supporting worktrees.
> 
> I've also tried the `pre-commit` way. Package does exist on Debian 12 and hooks do install, but then the hook doesn't work for me that way:
> 
> ```
> An error has occurred: InvalidManifestError: 
> ==> File /home/strk/.cache/pre-commit/repop4ji3b_1/.pre-commit-hooks.yaml
> ==> At Hook(id='clang-format')
> ==> At key: types_or
> ==> At index 9
> =====> Type tag 'textproto' is not recognized.  Try upgrading identify and pre-commit?
> Check the log at /home/strk/.cache/pre-commit/pre-commit.log
> ```
> 

Your python identify package is too old. Please see: https://github.com/pre-commit/pre-commit/issues/1946

> Is there any good reason for the increased complexity of using this `pre-commit` thing ?

It has already been discussed and decided. The QGIS ecosystem has moved on to pre-commit. Please see: https://github.com/qgis/QGIS-Enhancement-Proposals/issues/265 and https://github.com/qgis/QGIS/pull/59522

strk commented:
Can we at least give it as an option to contributors to EITHER use prepare-commit.sh OR pre-commit ?
It's being really hard for me:

python3-identify is already the newest version (2.5.17-1).

pipx upgrade identify
identify is already at latest version 2.6.6 (location: /home/strk/.local/pipx/venvs/identify)

```
git commit --amend
An error has occurred: InvalidManifestError: 
==> File /home/strk/.cache/pre-commit/repop4ji3b_1/.pre-commit-hooks.yaml
==> At Hook(id='clang-format')
==> At key: types_or
==> At index 9
=====> Type tag 'textproto' is not recognized.  Try upgrading identify and pre-commit?
```

I don't know why contributing to QGIS continues to become harder and harder.. what's the benefit of this ?

strk commented:
Is there any way to make the preparation mandatory when requirements are not met ?
Like prepare-commit.sh did just print a WARNING if `ag` was not present (for example) but didn't block you from committing, counting on CI to tell you if something was not as expected but at least giving you a chance to still check something locally and telling you how to check further ...

strk commented:
Using pre-commit version 4.1.0 fixed my issue. The apt package version was 3.0.4 -- I'll add a note about minimum required version too at this point, I guess ?

ptitjano commented:
> Using pre-commit version 4.1.0 fixed my issue. The apt package version was 3.0.4 -- I'll add a note about minimum required version too at this point, I guess ?

I think this is a good idea.
